discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,228 @@
+}
+
+func (conn *ZooKeeper) getNodeKey(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", conn.parentPath(service), service.ID)
+}
+
+func (conn *ZooKeeper) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, appName)
+}
+
+var zookeeperUpstreams = make(map[string][]ServiceNode)
+
+// CheckForUpstreamChanges checks another zookeeper node for changes
+func (conn *ZooKeeper) CheckForUpstreamChanges(backendName, backendTag string) bool {
+	// TODO: is there a way to filter by tag in zookeeper?
+	services, err := conn.getServices(backendName)
+	if err != nil {
+		log.Errorf("Failed to query %v: %s", backendName, err)
+		return false
+	}
+	didChange := zookeeperCompareForChange(zookeeperUpstreams[backendName], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		zookeeperUpstreams[backendName] = services
+	}
+	return didChange
+}
+
+func (conn *ZooKeeper) getServices(appName string) ([]ServiceNode, error) {
+	var services []ServiceNode
+
+	key := conn.getAppKey(appName)
+	children, _, error := conn.Client.Children(key)
+	if error != nil {
+		return services, error
+	}
+	for i := range children {
+		path := fmt.Sprintf("%s/%s", key, children[i])
+		data, _, error := conn.Client.Get(path)
+		if error != nil {
+			return services, error
+		}
+		srv, error := decodeZooKeeperNodeValue(data)
+		if error != nil {
+			return services, error
+		}
+		services = append(services, srv)
+	}
+	return services, nil
+}
+
+// Compare the two arrays to see if the address or port has changed
+// or if we've added or removed entries.
+func zookeeperCompareForChange(existing, new []ServiceNode) (changed bool) {
+	if len(existing) != len(new) {
+		return true
+	}
+
+	sort.Sort(ByZooKeeperServiceID(existing))
+	sort.Sort(ByZooKeeperServiceID(new))
+	for i, ex := range existing {
+		if ex.Address != new[i].Address ||
+			ex.Port != new[i].Port {
+			return true
+		}
+	}
+	return false
+}
+
+func (conn ZooKeeper) createParentPath(path string) error {
+	pathElements := strings.Split(path, "/")[1:]
+	sep := "/"
+	newPath := ""
+	for i := range pathElements {
+		newPath = strings.Join([]string{newPath, sep, pathElements[i]}, "")
+		if _, err := conn.Client.Create(newPath, nil, 0, zk.WorldACL(zk.PermAll)); err != nil && err != zk.ErrNodeExists {
+			return err
+		}
+	}
+	return nil
+}
+
+func (conn ZooKeeper) registerService(service *discovery.ServiceDefinition) error {
+	key := conn.getNodeKey(service)
+	if err := conn.createParentPath(conn.parentPath(service)); err != nil {
+		return err
+	}
+	value := encodeZooKeeperNodeValue(service)
+	if _, err := conn.Client.Create(
+		key,
+		[]byte(value),
+		zk.FlagEphemeral,
+		zk.WorldACL(zk.PermAll)); err != nil && err != zk.ErrNodeExists {
+		return err
+	}
+	_, _, ch, err := conn.Client.GetW(key)
```

Comment:
The edit swallowed the rest of my commentary, so circling back to the design question here...

@tgross wrote:
> I may be misunderstanding the code here -- can you explain the intent of what we're doing in this section? It looks like you're setting a watch on the key and then starting a new goroutine that creates a new watch on each event received. This seems like we're going to be firing up a new goroutine on every health check poll and then deregistering when the TTL expires but I can't figure out how we're passing that goroutine a message not to deregister.

@giorgio-v wrote:
> As for the checks, the watcher is set only if `k` does not exist. Does this clarify your doubts a bit?

I understand that, but what is the purpose of the watcher and how are we communicating with it?